### PR TITLE
[FLINK-6746] [docs] Updated Table API / SQL docs: Common API

### PR DESCRIPTION
**Pull request by fhueske**

### CODE REVIEW

File: docs/dev/table/common.md [hunk 1]
```diff
@@ -98,374 +89,767 @@ env.execute("Your Query")
...
-* Custom rules
+Apache Flink leverages Apache Calcite to optimize and translate queries. Among the optimization which are currently performed are projection and filter push-down, subquery decorrelation, and other kinds of query rewriting. Flink does not yet optimize the order of join but executes them in the same order as defined in the query (order of Tables in the `FROM` clause and/or order of join predicates in the `WHERE` clause).
+
+It is possible to tweak the set of optimization rules which are applied in different phases by providing a `CalciteConfig` object. This can be created via a builder by calling `CalciteConfig.createBuilder())` and provided to the TableEnvironment by calling `tableEnv.getConfig.setCalciteConfig(calciteConfig)`. 
 
 ### Explaining a Table
 
 The Table API provides a mechanism to explain the logical and optimized query plans to compute a `Table`. 
-This is done through the `TableEnvironment#explain(table)` method. It returns a string describing three plans: 
+This is done through the `TableEnvironment.explain(table)` method. It returns a String describing three plans:
```
Comment (sunjincheng121) [inline docs/dev/table/common.md]: I think we can improve the code format:
```
val table = table1
      .where('word.like("F%"))
      .unionAll(table2)
```
To
```
val table = table1
   .where('word.like("F%"))
   .unionAll(table2)
```
What do you think?

File: docs/dev/table/common.md [hunk 2]
```diff
@@ -98,374 +89,767 @@ env.execute("Your Query")
...
+
+// convert DataStream into Table with field names 'myName, 'myAge
+val table2: Table = tableEnv.fromDataStream(stream, 'name as 'myName, 'age as 'myAge)
+{% endhighlight %}
+</div>
+</div>
+
+#### Row
+
+The Row data type supports an arbitrary number of fields and fields with `null` values. Field names of rows can be specified via a `RowTypeInfo` or when converting a `Row` `DataStream` or `DataSet` into a `Table` (based on position).
```
Comment (alpinegizmo) [inline docs/dev/table/common.md]: Field names of rows can be => Field names can be

File: docs/dev/table/common.md [hunk 3]
```diff
@@ -98,374 +89,767 @@ env.execute("Your Query")
...
 {% top %}
 
+
 Query Optimization
 ------------------
 
-* No join order yet
-* Filter / Projection push down
-* Custom rules
+Apache Flink leverages Apache Calcite to optimize and translate queries. Among the optimization which are currently performed are projection and filter push-down, subquery decorrelation, and other kinds of query rewriting. Flink does not yet optimize the order of join but executes them in the same order as defined in the query (order of Tables in the `FROM` clause and/or order of join predicates in the `WHERE` clause).
```
Comment (alpinegizmo) [inline docs/dev/table/common.md]: ... The optimizations currently performed include projection and filter push-down, subquery decorrelation, and other kinds of query rewriting. Flink does not yet optimize the order of joins, but executes them in the same order as defined in the query ...

File: docs/dev/table/common.md [hunk 4]
```diff
@@ -98,374 +89,767 @@ env.execute("Your Query")
...
+
+It is possible to tweak the set of optimization rules which are applied in different phases by providing a `CalciteConfig` object. This can be created via a builder by calling `CalciteConfig.createBuilder())` and provided to the TableEnvironment by calling `tableEnv.getConfig.setCalciteConfig(calciteConfig)`.
```
Comment (alpinegizmo) [inline docs/dev/table/common.md]: ... and is provided to the TableEnvironment ...

File: docs/dev/table/common.md [hunk 5]
```diff
@@ -1,5 +1,5 @@
 ---
-title: "Concepts & Common API"
+title: "Common Concepts & API"
```
Comment (fhueske) [inline docs/dev/table/common.md]: Would "Joint" be a better choice?

File: docs/dev/table/common.md [hunk 6]
```diff
@@ -98,374 +89,767 @@ env.execute("Your Query")
...
+
 {% top %}
 
 Create a TableEnvironment
 -------------------------
 
-A `Table` is always bound to a specific `TableEnvironment`. It is not possible to combine Tables of different TableEnvironments.
+The `TableEnvironment` is a central concept of the Table API and SQL integration. It is responsible for:
+* Registering a `Table` in the internal catalog
+* Registering an external catalog
```
Comment (fhueske) [inline docs/dev/table/common.md]: It worked on my setup. Maybe some software to build the docs has a different version. I added a new line. Let me know if that fixes the issue

File: docs/dev/table/common.md [hunk 7]
```diff
@@ -98,374 +89,767 @@ env.execute("Your Query")
...
+
+A `TableEnvironment` is created by calling the static `TableEnvironment.getTableEnvironment()` method with a `StreamExecutionEnvironment` or an `ExecutionEnvironment` and an optional `TableConfig`. The `TableConfig` can be used to configure the `TableEnvironment` or to customize the query optimization and translation process (see [Query Optimization](#query-optimization)).
 
-**TODO: Extend**
+<div class="codetabs" markdown="1">
+<div data-lang="java" markdown="1">
+{% highlight java %}
+// ***************
+// STREAMING QUERY
+// ***************
```
Comment (fhueske) [inline docs/dev/table/common.md]: In most of the examples it says that the batch case is equivalent. The TableEnvironment is one of the few spots where there is an actual difference. I wanted to highlight that.

File: docs/dev/table/common.md [hunk 8]
```diff
@@ -98,374 +89,767 @@ env.execute("Your Query")
...
+// execute query
 {% endhighlight %}
 </div>
 
 <div data-lang="scala" markdown="1">
 {% highlight scala %}
-// works for StreamExecutionEnvironment identically
-val env = ExecutionEnvironment.getExecutionEnvironment
+// get a TableEnvironment
 val tableEnv = TableEnvironment.getTableEnvironment(env)
```
Comment (fhueske) [inline docs/dev/table/common.md]: No. In Scala there is no difference in the code because it infers the type.

File: docs/dev/table/common.md [hunk 9]
```diff
@@ -98,374 +89,767 @@ env.execute("Your Query")
...
-tableEnv.registerDataStream("Orders", ord, "user, product, amount");
+// register the ExternalCatalog catalog
+tableEnv.registerExternalCatalog("InMemCatalog", catalog);
 {% endhighlight %}
 </div>
 
 <div data-lang="scala" markdown="1">
 {% highlight scala %}
-val env = StreamExecutionEnvironment.getExecutionEnvironment
+// get a TableEnvironment
```
Comment (fhueske) [inline docs/dev/table/common.md]: No. In Scala there is no difference in the code because it infers the type.

File: docs/dev/table/common.md [hunk 10]
```diff
@@ -98,374 +89,767 @@ env.execute("Your Query")
...
 
-A `DataStream` is registered as a `Table` in a `StreamTableEnvironment` as follows:
+An external catalog can be created by implementing the `ExternalCatalog` interface and is registered in a `TableEnvironment` as follows:
 
 <div class="codetabs" markdown="1">
 <div data-lang="java" markdown="1">
 {% highlight java %}
-StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+// get a StreamTableEnvironment, works for BatchTableEnvironment equivalently
 StreamTableEnvironment tableEnv = TableEnvironment.getTableEnvironment(env);
```
Comment (fhueske) [inline docs/dev/table/common.md]: In Java, you have to specify the type of the returned TableEnvironment (StreamTableEnvironment for streaming and BatchTableEnvironment for batch)

File: docs/dev/table/common.md [hunk 11]
```diff
@@ -98,374 +89,767 @@ env.execute("Your Query")
...
+// register the DataStream as table "myTable2" with fields "myLong", "myString"
+tableEnv.registerDataStream("myTable2", stream, "myLong, myString");
+{% endhighlight %}
+</div>
 
-**TODO**
+<div data-lang="scala" markdown="1">
+{% highlight scala %}
+// get TableEnvironment. 
+// registration of a DataSet is equivalent
```
Comment (fhueske) [inline docs/dev/table/common.md]: In Scala a user is not necessarily aware that the TableEnvironment is a BatchTableEnvironment.

File: docs/dev/table/common.md [hunk 12]
```diff
@@ -98,374 +89,767 @@ env.execute("Your Query")
...
+There are two modes to convert a `Table` into a `DataStream`:
+
+1. **Append Mode**: This mode can only be used in the dynamic `Table` is only modified by `INSERT` changes, i.e, it is append-only and previously emitted results are never updated.
+2. **Retract Mode**: This mode can always be used. It encodes `INSERT` and `DELETE` changes with a `boolean` flag.
+
+<div class="codetabs" markdown="1">
+<div data-lang="java" markdown="1">
+{% highlight java %}
+// get StreamTableEnvironment. 
+StreamTableEnvironment tableEnv = TableEnvironment.getTableEnvironment(env);
```
Comment (fhueske) [inline docs/dev/table/common.md]: No, conversion of Table into DataSet is discussed below.

File: docs/dev/table/common.md [hunk 13]
```diff
@@ -98,374 +89,767 @@ env.execute("Your Query")
...
-A `Table` is always bound to a specific `TableEnvironment`. It is not possible to combine Tables of different TableEnvironments.
+The `TableEnvironment` is a central concept of the Table API and SQL integration. It is responsible for:
+* Registering a `Table` in the internal catalog
+* Registering an external catalog 
+* Executing SQL queries
+* Registering a user-defined (scalar, table, or aggregation) function
+* Converting a `DataStream` or `DataSet` into a `Table`
+* Holding a reference to an `ExecutionEnvironment` or `StreamExecutionEnvironment`
+
+A `Table` is always bound to a specific `TableEnvironment`. It is not process tables of different TableEnvironments in the same query, e.g., to join or union them.
```
Comment (fhueske) [inline docs/dev/table/common.md]: Yes, you are right

**Comment by fhueske** [comment]
Thanks for the reviews @sunjincheng121 and @alpinegizmo. I updated the PR.

File: docs/dev/table/common.md (hunk 6 above)
Comment (sunjincheng121) [inline docs/dev/table/common.md]: Yes, It's work well now.

**Comment by sunjincheng121** [comment]
Hi @fhueske thanks for the update. The PR looks good for me. 
+1 to merged.

**Comment by twalthr** [comment]
I will read this one more time and merge it.

**Comment by fhueske** [comment]
Thanks for merging @twalthr

**Comment by twalthr** [comment]
@fhueske can you close this PR?

**Pull request closed without merging**